### PR TITLE
Fix FreeFem PETSc install (REVIEW)

### PR DESCRIPTION
**Pull request by hajgato**
Needs https://github.com/easybuilders/easybuild-easyblocks/pull/1488

### CODE REVIEW

**Comment by hajgato** [comment]
Test fails on `magnycours`

**Comment by boegel** [comment]
@hajgato It fails without the patch on AMD Magny Cours, or with the patch? ;)

**Comment by hajgato** [comment]
@boegel What do you mean? In the previous version, the test was not performed. And now, the test passed on ivybridge, haswell, broadwell, and skylake (with the patch). I still waiting the new magnycours results, as a system library was missing. (libibverbs-devel). So maybe the test failed because of this.

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-easyblock = 'ConfigureMake'
+easyblock = 'EB_FreeFem'
```
Comment (boegel) [inline easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: This is not needed if you fix the class name for the easyblock in https://github.com/easybuilders/easybuild-easyblocks/pull/1488

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb [hunk 2]
```diff
@@ -13,26 +13,35 @@ toolchainopts = {'usempi': True, 'pic': True}
...
 dependencies = [
     ('HDF5', '1.10.1'),
     ('GSL', '2.4'),
     ('freeglut', '3.0.0'),
 ]
 
-preconfigopts = "autoreconf -i &&"
-
-configopts = '--with-hdf5="$EBROOTHDF5"/bin '
+configopts = '--with-hdf5="$EBROOTHDF5"/bin/h5pcc '
```
Comment (boegel) [inline easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: No need for the double qoutes around `$EBROOTHDF5`: `--with-hdf5=$EBROOTHDF5/bin/h5pcc` .

Also, why is this not handled by the `FreeFem++` easyblock in https://github.com/easybuilders/easybuild-easyblocks/pull/1488?

If there's an easyblock, the easyconfig should be as minimal as possible...

Same for rest of `configopts` & also `prebuildopts` below...

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# Pick up EB provided ScaLAPACK and FFTW in PETSc install 
+# August 23rd 2018 by B. Hajgato (Free University Brussels - VUB)
+--- freefem++-3.58/download/ff-petsc/Makefile.orig	2018-01-03 07:55:10.000000000 +0100
++++ freefem++-3.58/download/ff-petsc/Makefile	2018-08-25 13:21:53.409212134 +0200
+@@ -51,7 +51,7 @@
+ ifeq ($(MPICC),$(W_MPICC))
+   WITH_MPI_DIR=
+ else
+-  WITH_MPI_DIR="--with-mpi-dir='$(MPI_DIR)'"
++  WITH_MPI_DIR=--with-mpi-dir='$(MPI_DIR)'
```
Comment (boegel) [inline easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: @hajgato Why this change?

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+-	--with-blas-lapack-lib='$(LAPACKLIBS) $(BLASLIBS)' \
+-	--download-scalapack --download-metis --download-ptscotch \
++	--with-blas-lapack-lib="[${BLAS_LAPACK_LIB_DIR}/${BLAS_LAPACK_STATIC_LIBS}]" \
++	--with-scalapack=1 \
++	--with-scalapack-include=${SCALAPACK_INC_DIR} \
++	--with-scalapack-lib="[${SCALAPACK_LIB_DIR}/${SCALAPACK_STATIC_LIBS}]" \
++	--download-metis --download-ptscotch \
+ 	--download-mumps --download-ml --download-hypre --download-parmetis \
+-	--download-superlu --download-pastix --download-suitesparse --download-fftw \
++	--download-superlu --download-suitesparse \
```
Comment (boegel) [inline easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: Why drop `--download-pastix`?

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-easyblock = 'ConfigureMake'
+easyblock = 'EB_FreeFem_plus__plus_'
```
Comment (boegel) [inline easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: @hajgato This should not be needed, will be auto-derived based on software name...

Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: Done

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb (hunk 2 above)
Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: @boegel Do you want me to put the `runtest`, `sanity_check_path`, and `modextrapaths` as well in the easyblock?

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch (hunk 3 above)
Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: Beacuse without it the build failed. I did not look into the details. But does it really matters?

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb (hunk 2 above)
Comment (boegel) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: Yes, that makes most sense to me...

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch (hunk 3 above)
Comment (boegel) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: It's weird that it's needed, I was just wondering why that change was made. I guess it doesn't matter too much as long as it fixes a problem. ;)

Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: I think it is the nature of `PETSc` build, it is a `Python` build system. If I run the `PETSc` installation script by hand (outside `EasyBuild`), it works with the double-quotation marks, using `EasyBuild` I had to remove it. (`PETSc` script was complaining something like cannot configure with `--with-mpi-dir` value.) So probably because it is a `Python` in a `Python`

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch (hunk 4 above)
Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58_ebvars.patch]: `PaStiX` does not work parallel, so it is disabled in the configuration (see now the easyblock). Here, it also breaks the `PETSc` parallel build.

File: easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb (hunk 2 above)
Comment (hajgato) [reply easybuild/easyconfigs/f/FreeFem++/FreeFem++-3.58-foss-2017b-downloaded-deps.eb]: OK.

**Comment by hajgato** [comment]
3 out of 291 tests were failed on `Magny-cours` architecture (AMD Opteron(tm) Processor 6134)). All tests were passed on Intel `Ivybrige (v2)`, `Haswell (v3)`, `Broadwell (v4)` and `Skylake (gold)` architectures. Therefore, I removed the test from the easyblock, so it could be easier put back to the easyconfig. Since `Magny-Cours` is from 2010, I did not spend any time to figure out the reason for the failed tests. All the 3 failed test had the error: `mshptg bug 2` and then points are too close, so I guess it should be some numeric instability.

**Comment by boegel** [comment]
I'm closing this in favor of more recent FreeFEM versions, see #9902 for example.

The `foss/2017b` toolchain is pretty ancient by now, and the easyblock PR that this requires has been superseded by another (see https://github.com/easybuilders/easybuild-easyblocks/pull/1969) which may not work with FreeFEM 3.x.

Not worth the effort to pursue this imho, sorry @hajgato ...
